sql/core/src/main/scala/org/apache/spark/sql/execution/ui/AllExecutionsPage.scala

```diff
@@ -121,65 +122,247 @@ private[ui] class AllExecutionsPage(parent: SQLTab) extends WebUIPage("") with L
+    pageSize: Int,
+    sortColumn: String,
+    desc: Boolean,
     showRunningJobs: Boolean,
     showSucceededJobs: Boolean,
-    showFailedJobs: Boolean) {
+    showFailedJobs: Boolean) extends PagedTable[ExecutionTableRowData] {
 
-  protected def baseHeader: Seq[String] = Seq(
-    "ID",
-    "Description",
-    "Submitted",
-    "Duration")
+  override val dataSource = new ExecutionDataSource(
+    request,
+    parent,
+    data,
+    basePath,
+    currentTime,
+    pageSize,
+    sortColumn,
+    desc)
 
-  protected def header: Seq[String]
+  val parameterPath = basePath + s"/$subPath/?" + parameterOtherTable.mkString("&")
 
-  protected def row(
-      request: HttpServletRequest,
-      currentTime: Long,
-      executionUIData: SQLExecutionUIData): Seq[Node] = {
-    val submissionTime = executionUIData.submissionTime
-    val duration = executionUIData.completionTime.map(_.getTime()).getOrElse(currentTime) -
-      submissionTime
+  override def tableId: String = executionTag + "-table"
 
-    def jobLinks(status: JobExecutionStatus): Seq[Node] = {
-      executionUIData.jobs.flatMap { case (jobId, jobStatus) =>
-        if (jobStatus == status) {
-          <a href={jobURL(request, jobId)}>[{jobId.toString}]</a>
+  override def tableCssClass: String =
+    "table table-bordered table-condensed table-striped " +
+      "table-head-clickable table-cell-width-limited"
+
+  override def prevPageSizeFormField: String = executionTag + ".prevPageSize"
+
+  override def pageLink(page: Int): String = {
+    val encodedSortColumn = URLEncoder.encode(sortColumn, "UTF-8")
+    parameterPath +
+      s"&$pageNumberFormField=$page" +
+      s"&$executionTag.sort=$encodedSortColumn" +
+      s"&$executionTag.desc=$desc" +
+      s"&$pageSizeFormField=$pageSize" +
+      s"#$tableHeaderId"
+  }
+
+  override def pageSizeFormField: String = executionTag + ".pageSize"
+
+  override def pageNumberFormField: String = executionTag + ".page"
+
+  override def goButtonFormPath: String = {
+    val encodedSortColumn = URLEncoder.encode(sortColumn, "UTF-8")
+    s"$parameterPath&$executionTag.sort=$encodedSortColumn&$executionTag.desc=$desc#$tableHeaderId"
+  }
+
+  override def headers: Seq[Node] = {
+    // Information for each header: title, cssClass, and sortable
+    val executionHeadersAndCssClasses: Seq[(String, String, Boolean)] =
+      Seq(("ID", "", true), ("Description", "", true), ("Submitted", "", true),
+        ("Duration", "", true)) ++ {
+        if (showRunningJobs && showSucceededJobs && showFailedJobs) {
+          Seq(("Running Job IDs", "", true), ("Succeeded Job IDs", "", true),
+            ("Failed Job IDs", "", true))
+        } else if (showSucceededJobs && showFailedJobs) {
+          Seq(("Succeeded Job IDs", "", true), ("Failed Job IDs", "", true))
+        }
+        else {
+          Seq(("Job IDs", "", true))
+        }
+      }
+
+    if (!executionHeadersAndCssClasses.filter(_._3).map(_._1).contains(sortColumn)) {
```

Comment:
This might be worth using case expressions to clarify what _3 and _1 are.

Reply:
Thanks. Done.